@jbaudoux one test is failing on `odoo/odoo`
```python
FAIL: LocationContentTransferSetDestinationAllCase.test_set_destination_all_with_partially_available_move_with_ancestor
Traceback (most recent call last):
  File "/home/travis/build/OCA/wms/setup/shopfloor/odoo/addons/shopfloor/tests/test_location_content_transfer_set_destination_all.py", line 161, in test_set_destination_all_with_partially_available_move_with_ancestor
    params={"location_id": self.content_loc.id, "barcode": sub_shelf1.barcode},
  File "/home/travis/build/OCA/wms/setup/shopfloor_base/odoo/addons/shopfloor_base/services/service.py", line 27, in dispatch
    return super().dispatch(method_name, *args, params=params)
  File "/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/odoo/addons/rest_log/components/service.py", line 36, in dispatch
    return super().dispatch(method_name, *args, params=params)
  File "/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/odoo/addons/base_rest/components/service.py", line 158, in dispatch
    res = method(*args, **secure_params)
  File "/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/odoo/addons/base_rest/restapi.py", line 61, in response_wrap
    response = f(*args, **kw)
  File "/home/travis/build/OCA/wms/setup/shopfloor/odoo/addons/shopfloor/services/location_content_transfer.py", line 477, in set_destination_all
    self._set_all_destination_lines_and_done(pickings, move_lines, scanned_location)
  File "/home/travis/build/OCA/wms/setup/shopfloor/odoo/addons/shopfloor/services/location_content_transfer.py", line 434, in _set_all_destination_lines_and_done
    stock.validate_moves(move_lines.move_id)
  File "/home/travis/build/OCA/wms/setup/shopfloor/odoo/addons/shopfloor/actions/stock.py", line 30, in validate_moves
    moves_todo.extract_and_action_done()
  File "/home/travis/build/OCA/wms/setup/shopfloor/odoo/addons/shopfloor/models/stock_move.py", line 104, in extract_and_action_done
    assert new_picking.state == "assigned"
AssertionError
```
I restart the Travis job to check as all was green on `OCA/OCB`...

--- a/stock_dynamic_routing/README.rst
+++ b/stock_dynamic_routing/README.rst
@@ -156,12 +156,14 @@ Authors
 ~~~~~~~
 
 * Camptocamp
+* BCIM
 
 Contributors
 ~~~~~~~~~~~~
 
 * Joël Grand-Guillaume <joel.grandguillaume@camptocamp.com>
 * Guewen Baconnier <guewen.baconnier@camptocamp.com>
+* Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
 * Akim Juillerat <akim.juillerat@camptocamp.com>
 
 Maintainers

--- a/stock_dynamic_routing/models/stock_move.py
+++ b/stock_dynamic_routing/models/stock_move.py
@@ -1,7 +1,8 @@
 # Copyright 2019-2020 Camptocamp (https://www.camptocamp.com)
+# Copyright 2021 Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
 import uuid
-from collections import defaultdict, namedtuple
+from collections import OrderedDict, defaultdict, namedtuple
 
 from psycopg2 import sql
 
@@ -236,13 +237,27 @@ class StockMove(models.Model):
         """
         pickings_to_check_for_emptiness = self.env["stock.picking"]
         move_ids_to_assign_per_location = defaultdict(list)
-        for move in self:
-            routing_rule = routing_details[move].rule
-            if not routing_rule.method == "pull":
+        move_ids_to_assign_nonrelocated = []
+        next_moves_to_update = self.browse()
+        routing_to_apply = [
+            (move, detail.rule) for move, detail in routing_details.items()
+        ]
+        for move, routing_rule in routing_to_apply:
+            if routing_rule.method == "push":
+                # In this case, we should not assign the move inside the
+                # pull dynamic routing. The push move must first be added before the
+                # initial move is assigned. Otherwise, when the destination
+                # location of the initial move is changed by the push rule, the
+                # putaway won't be recomputed as it is already assigned.
+                continue
+
+            if not routing_rule:
+                move_ids_to_assign_nonrelocated.append(move.id)
                 continue
 
             if move.picking_id.picking_type_id == routing_rule.picking_type_id:
                 # already correct
+                move_ids_to_assign_nonrelocated.append(move.id)
                 continue
 
             # we expect all the lines to go to the same destination for
@@ -256,21 +271,39 @@ class StockMove(models.Model):
             dest_location = move.location_dest_id
             rule_location = routing_rule.location_dest_id
             if rule_location.is_sublocation_of(dest_location):
-                # The destination of the move, as a parent of the destination
+                # The destination of the move, is a parent of the destination
                 # of the routing, goes to the correct place, but is not precise
                 # enough: set the new destination to match the rule's one.
-                # The source of the dest. move will be changed to match it,
-                # which may reapply a new routing rule on the dest. move.
-                move._routing_pull_switch_destination(routing_rule)
+                move.location_dest_id = routing_rule.location_dest_id
+                # After the reservation, the source of the dest. move will be
+                # changed to match it, which may reapply a new routing rule on
+                # the dest. move.
+                next_moves_to_update |= move.move_dest_ids.filtered(
+                    lambda r: r.state == "waiting"
+                )
 
             elif not dest_location.is_sublocation_of(rule_location):
                 # The destination of the move is unrelated (nor identical, nor
                 # a parent or a child) to the routing destination: in this case
-                # we have to add a routing move before the current move to
-                # route the goods in the correct place
-                move._routing_pull_insert_move(
-                    routing_rule, current_picking_type, original_destination
+                # we have to add a routing move after to reach the original destination
+                move.location_dest_id = routing_rule.location_dest_id
+                # create a copy of the move with the current picking type and
+                # going to its original destination: it will be assigned to the
+                # same picking as the original picking of our move
+                routing_move = move._insert_routing_moves(
+                    current_picking_type,
+                    # the source of the next move has to be the same as the
+                    # destination of the current move
+                    routing_rule.location_dest_id,
+                    original_destination,
                 )
+                # Get any routing for this new move
+                routing_rule = self.env["stock.routing"]._routing_rule_for_moves(
+                    routing_move
+                )[routing_move]
+                if routing_rule:
+                    # Add a new routing to apply for this new move
+                    routing_to_apply.append((routing_move, routing_rule))
 
             pickings_to_check_for_emptiness |= move.picking_id
             move._assign_picking()
@@ -316,60 +349,52 @@ class StockMove(models.Model):
         # by another move, sort the moves by their source location, from the
         # most precise to the least precise. The order of the move ids within
         # one location is preserved.
+        #
+        # The non routed moves are assigned at last. This allows compatibility
+        # with the module stock_move_source_relocate and ensures we call
+        # _action_assign on the complete set of moves
         sorted_locations = sorted(
             move_ids_to_assign_per_location, key=lambda l: l.parent_path, reverse=True
         )
         to_assign_ids = []
         for location in sorted_locations:
             to_assign_ids += move_ids_to_assign_per_location[location]
+        to_assign_ids += move_ids_to_assign_nonrelocated
 
         self.browse(to_assign_ids).with_context(
             exclude_apply_dynamic_routing=True
         )._action_assign()
 
+        # Update the destination moves. This might trigger a new routing on the
+        # destination move.
+        next_moves_to_update._routing_pull_switch_source()
+
         pickings_to_check_for_emptiness._dynamic_routing_handle_empty()
 
-    def _routing_pull_switch_destination(self, routing_rule):
-        """Switch the destination of the move in place
+    def _routing_pull_switch_source(self):
+        """Switch the source location of the move in place.
 
-        In this case, do not insert a new move but switch the destination
-        of the current move. The destination move source location is changed
-        as well, which might trigger a new routing on the destination move.
+        When the destination of the next move has changed, we do not insert a
+        new move but switch the source location of the current move.  This
+        might trigger a new routing on the destination move.
         """
-        self.location_dest_id = routing_rule.location_dest_id
-        next_move = self.move_dest_ids.filtered(lambda r: r.state == "waiting")
-        # FIXME what should happen if we have > 1 move? What would
-        # be the use case?
-        if next_move and len(next_move) == 1:
-            split_move = self.browse(next_move._split(self.product_qty))
-            if split_move != next_move:
-                # No split occurs if the quantity was the same.
-                # But if it did split, detach it from the move on which
-                # we have a different routing
-                next_move.move_orig_ids -= self
-                split_move.move_orig_ids = self
-            next_move = split_move
-
-        next_move.location_id = routing_rule.location_dest_id
-
-    def _routing_pull_insert_move(self, routing_rule, picking_type, destination):
-        """Add a move after the current move to reach the destination
-
-        The routing rules are applied on the new move in case it would trigger
-        a new move or a switch of picking type.
-        """
-        self.location_dest_id = routing_rule.location_dest_id
-        # create a copy of the move with the current picking type and
-        # going to its original destination: it will be assigned to the
-        # same picking as the original picking of our move
-        routing_move = self._insert_routing_moves(
-            picking_type,
-            # the source of the next move has to be the same as the
-            # destination of the current move
-            routing_rule.location_dest_id,
-            destination,
-        )
-        routing_move._chain_apply_routing()
+        for move in self:
+            origmoves_by_location = OrderedDict()
+            for orig_move in move.move_orig_ids:
+                origmoves_by_location.setdefault(
+                    orig_move.location_dest_id, move.browse()
+                )
+                origmoves_by_location[orig_move.location_dest_id] |= orig_move
+            for location_id, orig_moves in reversed(origmoves_by_location.items()):
+                qty = sum(orig_moves.mapped("product_qty"))
+                split_move = self.browse(move._split(qty))
+                if split_move != move:
+                    # No split occurs if the quantity was the same.
+                    # But if it did split, detach it from the move on which
+                    # we have a different routing
+                    move.move_orig_ids -= orig_moves
+                    split_move.move_orig_ids = orig_moves
+                split_move.location_id = location_id
 
     def _apply_routing_rule_push(self, routing_details):
         """Apply push dynamic routing

--- a/stock_dynamic_routing/readme/CONTRIBUTORS.rst
+++ b/stock_dynamic_routing/readme/CONTRIBUTORS.rst
@@ -1,3 +1,4 @@
 * Joël Grand-Guillaume <joel.grandguillaume@camptocamp.com>
 * Guewen Baconnier <guewen.baconnier@camptocamp.com>
+* Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
 * Akim Juillerat <akim.juillerat@camptocamp.com>

--- a/stock_dynamic_routing/tests/test_routing_pull.py
+++ b/stock_dynamic_routing/tests/test_routing_pull.py
@@ -115,6 +115,7 @@ class TestRoutingPullCommon(common.SavepointCase):
                     "product_uom_qty": qty,
                     "product_uom": product.uom_id.id,
                     "picking_id": customer_picking.id,
+                    "picking_type_id": wh.out_type_id.id,
                     "location_id": wh.wh_output_stock_loc_id.id,
                     "location_dest_id": self.customer_loc.id,
                     "state": "waiting",
@@ -129,6 +130,7 @@ class TestRoutingPullCommon(common.SavepointCase):
                     "product_uom_qty": qty,
                     "product_uom": product.uom_id.id,
                     "picking_id": pick_picking.id,
+                    "picking_type_id": wh.pick_type_id.id,
                     "location_id": wh.lot_stock_id.id,
                     "location_dest_id": wh.wh_output_stock_loc_id.id,
                     "move_dest_ids": [(4, dest.id)],
@@ -785,6 +787,7 @@ class TestRoutingPull(TestRoutingPullCommon):
             move_a | move_b | move_c | move_d,
             [
                 {
+                    "product_qty": 4,
                     "move_orig_ids": [],
                     "move_dest_ids": move_b.ids,
                     "state": "confirmed",
@@ -792,6 +795,7 @@ class TestRoutingPull(TestRoutingPullCommon):
                     "location_dest_id": self.wh.wh_output_stock_loc_id.id,
                 },
                 {
+                    "product_qty": 4,
                     "move_orig_ids": move_a.ids,
                     "move_dest_ids": [],
                     "state": "waiting",
@@ -799,6 +803,7 @@ class TestRoutingPull(TestRoutingPullCommon):
                     "location_dest_id": self.customer_loc.id,
                 },
                 {
+                    "product_qty": 6,
                     "move_orig_ids": [],
                     "move_dest_ids": move_d.ids,
                     "state": "assigned",
@@ -806,6 +811,7 @@ class TestRoutingPull(TestRoutingPullCommon):
                     "location_dest_id": area1.id,
                 },
                 {
+                    "product_qty": 6,
                     "move_orig_ids": move_c.ids,
                     "move_dest_ids": [],
                     "state": "waiting",

--- a/stock_move_source_relocate_dynamic_routing/tests/__init__.py
+++ b/stock_move_source_relocate_dynamic_routing/tests/__init__.py
@@ -1,1 +1,2 @@
 from . import test_dynamic_relocate
+from . import test_dynamic_destination

--- a/stock_move_source_relocate_dynamic_routing/tests/test_dynamic_destination.py
+++ b/stock_move_source_relocate_dynamic_routing/tests/test_dynamic_destination.py
@@ -1,0 +1,254 @@
+from odoo.addons.stock_dynamic_routing.tests.test_routing_pull import (
+    TestRoutingPullCommon,
+)
+
+
+class TestRoutingAndSourceRelocate(TestRoutingPullCommon):
+    def test_pickship_with_routing_and_relocate(self):
+        """Check that if a pick move is partially available, the ship move is not split
+
+        When a pick move is partially available and the destination is
+        re-routed, it is split in 2 moves, one assigned on the re-routed
+        destination location and one not assigned relocated and then re-routed
+        to that same destination location. As both pick moves are going to the
+        same destination location, ensure the ship move is not split.
+
+        This test is similar to test_change_dest_move_source_split in
+        stock_dynamic_routing/tests/test_routing_pull except that no move_d is created"""
+
+        self.env["stock.source.relocate"].create(
+            {
+                "location_id": self.wh.lot_stock_id.id,
+                "picking_type_id": self.wh.pick_type_id.id,
+                "relocate_location_id": self.location_hb.id,
+                "rule_domain": "[]",
+            }
+        )
+
+        area1 = self.env["stock.location"].create(
+            {"location_id": self.wh.wh_output_stock_loc_id.id, "name": "Area1"}
+        )
+        self.pick_type_routing_op.default_location_dest_id = area1
+
+        pick_type_routing_delivery = self.env["stock.picking.type"].create(
+            {
+                "name": "Delivery (after routing)",
+                "code": "outgoing",
+                "sequence_code": "OUT(R)",
+                "warehouse_id": self.wh.id,
+                "use_create_lots": False,
+                "use_existing_lots": True,
+                "default_location_src_id": area1.id,
+                "default_location_dest_id": self.customer_loc.id,
+            }
+        )
+        self.env["stock.routing"].create(
+            {
+                "location_id": area1.id,
+                "picking_type_id": self.wh.out_type_id.id,
+                "rule_ids": [
+                    (
+                        0,
+                        0,
+                        {
+                            "method": "pull",
+                            "picking_type_id": pick_type_routing_delivery.id,
+                        },
+                    )
+                ],
+            }
+        )
+
+        pick_picking, customer_picking = self._create_pick_ship(
+            self.wh, [(self.product1, 10)]
+        )
+        move_a = pick_picking.move_lines
+        move_b = customer_picking.move_lines
+
+        self._update_product_qty_in_location(self.location_hb_1_2, move_a.product_id, 6)
+        pick_picking.action_assign()
+
+        move_c = (
+            self.env["stock.picking"]
+            .search([("picking_type_id", "=", self.pick_type_routing_op.id)])
+            .move_lines
+        )[0]
+        self.assertRecordValues(
+            move_a | move_b | move_c,
+            [
+                {
+                    "product_qty": 4,
+                    "move_orig_ids": [],
+                    "move_dest_ids": move_b.ids,
+                    "state": "confirmed",
+                    "location_id": self.location_hb.id,
+                    "location_dest_id": area1.id,
+                },
+                {
+                    "product_qty": 10,
+                    "move_orig_ids": (move_c | move_a).ids,
+                    "move_dest_ids": [],
+                    "state": "waiting",
+                    "location_id": area1.id,
+                    "location_dest_id": self.customer_loc.id,
+                },
+                {
+                    "product_qty": 6,
+                    "move_orig_ids": [],
+                    "move_dest_ids": move_b.ids,
+                    "state": "assigned",
+                    "location_id": self.location_hb.id,
+                    "location_dest_id": area1.id,
+                },
+            ],
+        )
+
+        self.assertEqual(move_a.picking_id.picking_type_id, self.pick_type_routing_op)
+        self.assertEqual(move_b.picking_id.picking_type_id, pick_type_routing_delivery)
+        self.assertEqual(move_c.picking_id.picking_type_id, self.pick_type_routing_op)
+
+    def test_prepickship_with_routing_and_relocate(self):
+        """Check that if a pick move is waiting another move, the ship move is not split
+
+        When a pick move is partially available with a new additional pre-pick
+        step and the destination is re-routed, it is split in 2 moves, one
+        waiting the new move on the re-routed destination location and one not
+        assigned relocated and then re-routed to that same destination
+        location. As both pick moves are going to the same destination
+        location, ensure the ship move is not split.
+
+        This test is similar to test_change_dest_move_source_split and
+        test_change_dest_move_source_chain in
+        stock_dynamic_routing/tests/test_routing_pull except that no move_d is
+        created"""
+
+        self.env["stock.source.relocate"].create(
+            {
+                "location_id": self.wh.lot_stock_id.id,
+                "picking_type_id": self.wh.pick_type_id.id,
+                "relocate_location_id": self.location_hb.id,
+                "rule_domain": "[]",
+            }
+        )
+
+        # Configure Pre-Pick routing
+        location_shu = self.env["stock.location"].create(
+            {"location_id": self.location_hb.id, "name": "SHU"}
+        )
+        prepick_pick_type = self.env["stock.picking.type"].create(
+            {
+                "name": "PrePick",
+                "code": "internal",
+                "sequence_code": "WH/SHU",
+                "warehouse_id": self.wh.id,
+                "use_create_lots": False,
+                "use_existing_lots": True,
+                "default_location_src_id": location_shu.id,
+                "default_location_dest_id": self.location_hb.id,
+            }
+        )
+        self.env["stock.routing"].create(
+            {
+                "location_id": location_shu.id,
+                "picking_type_id": self.wh.pick_type_id.id,
+                "rule_ids": [
+                    (0, 0, {"method": "pull", "picking_type_id": prepick_pick_type.id})
+                ],
+            }
+        )
+
+        # Configure Pick-Ship routing
+        area1 = self.env["stock.location"].create(
+            {"location_id": self.wh.wh_output_stock_loc_id.id, "name": "Area1"}
+        )
+        self.pick_type_routing_op.default_location_dest_id = area1
+
+        pick_type_routing_delivery = self.env["stock.picking.type"].create(
+            {
+                "name": "Delivery (after routing)",
+                "code": "outgoing",
+                "sequence_code": "OUT(R)",
+                "warehouse_id": self.wh.id,
+                "use_create_lots": False,
+                "use_existing_lots": True,
+                "default_location_src_id": area1.id,
+                "default_location_dest_id": self.customer_loc.id,
+            }
+        )
+        self.env["stock.routing"].create(
+            {
+                "location_id": area1.id,
+                "picking_type_id": self.wh.out_type_id.id,
+                "rule_ids": [
+                    (
+                        0,
+                        0,
+                        {
+                            "method": "pull",
+                            "picking_type_id": pick_type_routing_delivery.id,
+                        },
+                    )
+                ],
+            }
+        )
+
+        pick_picking, customer_picking = self._create_pick_ship(
+            self.wh, [(self.product1, 10)]
+        )
+        move_a = pick_picking.move_lines
+        move_b = customer_picking.move_lines
+
+        self._update_product_qty_in_location(location_shu, move_a.product_id, 6)
+        pick_picking.action_assign()
+
+        move_d = (
+            self.env["stock.picking"]
+            .search([("picking_type_id", "=", prepick_pick_type.id)])
+            .move_lines
+        )
+        move_c = (
+            self.env["stock.picking"]
+            .search([("picking_type_id", "=", self.pick_type_routing_op.id)])
+            .move_lines
+        )[0]
+        self.assertRecordValues(
+            move_a | move_b | move_c | move_d,
+            [
+                {
+                    "product_qty": 4,
+                    "move_orig_ids": [],
+                    "move_dest_ids": move_b.ids,
+                    "state": "confirmed",
+                    "location_id": self.location_hb.id,
+                    "location_dest_id": area1.id,
+                },
+                {
+                    "product_qty": 10,
+                    "move_orig_ids": (move_c | move_a).ids,
+                    "move_dest_ids": [],
+                    "state": "waiting",
+                    "location_id": area1.id,
+                    "location_dest_id": self.customer_loc.id,
+                },
+                {
+                    "product_qty": 6,
+                    "move_orig_ids": move_d.ids,
+                    "move_dest_ids": move_b.ids,
+                    "state": "waiting",
+                    "location_id": self.location_hb.id,
+                    "location_dest_id": area1.id,
+                },
+                {
+                    "product_qty": 6,
+                    "move_orig_ids": [],
+                    "move_dest_ids": move_c.ids,
+                    "state": "assigned",
+                    "location_id": location_shu.id,
+                    "location_dest_id": self.location_hb.id,
+                },
+            ],
+        )
+
+        self.assertEqual(move_a.picking_id.picking_type_id, self.pick_type_routing_op)
+        self.assertEqual(move_b.picking_id.picking_type_id, pick_type_routing_delivery)
+        self.assertEqual(move_c.picking_id.picking_type_id, self.pick_type_routing_op)

